stellargraph/core/element_data.py

```diff
@@ -446,16 +589,23 @@ def degrees(self, *, ins=True, outs=True):
         adj = self._adj_lookup(ins=ins, outs=outs)
         return defaultdict(int, ((key, len(value)) for key, value in adj.items()))
 
-    def edge_ilocs(self, node_id, *, ins, outs) -> np.ndarray:
+    def edge_ilocs(
+        self, node_iloc, *, ins, outs, other_node_type_iloc=None
+    ) -> np.ndarray:
         """
-        Return the integer locations of the edges for the given node_id
+        Return the integer locations of the edges for the given node iloc
 
         Args:
-            node_id: the ID of the node
+            node_iloc: the iloc of the node
 
 
         Returns:
-            The integer locations of the edges for the given node_id.
+            The integer locations of the edges for the given node iloc.
         """
 
-        return self._adj_lookup(ins=ins, outs=outs)[node_id]
+        if other_node_type_iloc is not None:
+            return self._adj_lookup_by_other_node_type(ins=ins, outs=outs).get(
+                other_node_type_iloc, defaultdict(list)
+            )[node_iloc]
```

Comment:
I take it the `get`+`defaultdict(list)` here is in case there's node types that don't have any incoming edges or any outgoing edges (e.g. all edges might be `a->b`)? I'm slightly concerned that creating a `defaultdict` for every call may be adding noticeable overhead, and, also it's returning a `list` rather than a NumPy array, so consumers may get confused (and/or have to spend more work converting types).



One option might be pre-filling the `_edges..._dict_by_other_node_type` values with every node type, e.g.:

```python
class EmptyFlatAdjacencyList:
    def __init__(self, dtype):
        self.empty = np.array([], dtype=dtype)

    def __getitem__(self, idx):
        return empty
```

Or just creating an O(nodes) `FlatAdjacencyList`, like `FlatAdjacencyList(np.array([], dtype=...), np.zeros(len(nodes) + 1))`

Another option would be moving the defaulting to a separate code path, like:

```python
adj_list = self._adj_lookup_by_other_node_type(ins=ins, outs=outs).get(other_node_type_iloc)
if adj_list is None:
    return np.array([], dtype=...)
return adj_list[node_iloc]
```

Reply:
Ah good point, I did want to also allow completely non-existent node types to return empty lists:
```python
g = StellarGraph(...)  # has node types ["A", "B"]
g.neighbors(1, other_node_type="X")  # should return np.array([])
```
So I might go with the separate code path approach?

Reply:
What's the motivation for returning empty, rather than emitting an error?